src/compiler/checker.ts

```diff
@@ -10415,6 +10415,11 @@ module ts {
                 });
             }
 
+            // Non-ambient classes cannot merge with interfaces.
+            if (!(node.flags & NodeFlags.Ambient) && getDeclarationOfKind(symbol, SyntaxKind.InterfaceDeclaration)) {

```

Comment:
Need to check if the node is in an ambient context. You can use getEffectiveDeclarationFlags if you want